lightwood/mixers/helpers/ar_net.py

```diff
@@ -0,0 +1,62 @@
+import torch
+from torch import nn
+from lightwood.mixers.helpers.default_net import DefaultNet
+from lightwood.helpers.torch import LightwoodAutocast
+from itertools import chain
+
+
+class ArNet(DefaultNet):
+    """
+    DefaultNet variant that adds a secondary stream (simple linear layer) with constrained
+    weights to learn autoregressive coefficients for numerical time series targets
+    """
+    def __init__(self, dynamic_parameters,
+                 transformer,
+                 input_size=None,
+                 output_size=None,
+                 nr_outputs=None,
+                 shape=None,
+                 max_params=3e5,
+                 dropout=None,
+                 pretrained_net=None):
+        """
+        :param data_info: lightwood.mixers.helpers.transformer.Transformer;
+        used to save the indexes for which we learn the autoregressive branch of the network
+        """
+        assert len(transformer.output_features) == 1  # ArNet supports single target prediction only
+        super().__init__(dynamic_parameters,
+                         input_size=input_size,
+                         output_size=output_size,
+                         nr_outputs=nr_outputs,
+                         shape=shape,
+                         max_params=max_params,
+                         dropout=dropout,
+                         pretrained_net=pretrained_net
+                         )
+        target = transformer.output_features[0]
+        self.transformer = transformer
+        self.ar_column = f'__mdb_ts_previous_{target}'
+        self.ar_idxs = list(*[transformer.input_idxs[col]
+                              for col in transformer.input_idxs
+                              if col == self.ar_column])
+
+        dims = [(len(self.ar_idxs), transformer.feature_len_map[target])]
+        linears = [nn.Linear(in_features=inf, out_features=outf) for inf, outf in dims]
+        self.ar_net = nn.Sequential(*linears)
+        self.ar_net.to(self.device)
+
+    def to(self, device=None, available_devices=None):
+        self.ar_net.to(device)
+        return super().to(device, available_devices)
+
+    def forward(self, input):
+        with LightwoodAutocast():
+            residual_output = self._foward_net(input)
+            ar_output = self.ar_net(input[:, self.ar_idxs])
```

Comment:
This is a cool approach - how do the weight updates work here? what I'm seeing is currently:

output1 = net1(inp)
output2 = net2(inp)

but the loss would combine output1 + output2 (as the forward pass sums the two terms up). 

I'm curious to know what happens when dynamics of net1/net2 differ (i.e. output1 >>> output2 or vice versa).

Reply:
I'll admit this is an analysis I did not make. In theory outputs should be in the same ballpark as they're both regularized (in part why I chose such a big decay). The original inspiration for the AR stream in this network is based off two papers:

1) https://arxiv.org/abs/1911.12436 ; for the actual layer definition
2) https://arxiv.org/pdf/1703.07015.pdf ; for the combined final output + loss propagation over two streams. See section 4.6 for some interesting analysis, though it doesn't exactly address your question. I guess we could further explore this at some point!